# 1. Importar librerías

In [62]:
#Librerías para manipulación de datos
import pandas as pd

#Librerías para visualización
import matplotlib.pyplot as plt
import seaborn as sns
from plotly.subplots import make_subplots
import plotly.express as px
import plotly.graph_objects as go

#Importar módulo de funciones
import sys
sys.path.insert(0, 'utils')  # Añadir el directorio 'utils' al PATH
import funciones
from importlib import reload
reload(funciones)  # Recarga el módulo

<module 'funciones' from 'c:\\Users\\opaac\\OneDrive - Universidad de Antioquia\\UdeA - SSM\\2024-1\\Analítica\\recursos_humanos\\utils\\funciones.py'>

# 2. Importar datos

In [3]:
data = funciones.get_data('data_2015')

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4410 entries, 0 to 4409
Data columns (total 30 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   InfoDate                 4410 non-null   dbdate 
 1   EmployeeID               4410 non-null   Int64  
 2   Age                      4410 non-null   Int64  
 3   DistanceFromHome         4410 non-null   Int64  
 4   MonthlyIncome            4410 non-null   Int64  
 5   NumCompaniesWorked       4410 non-null   float64
 6   PercentSalaryHike        4410 non-null   Int64  
 7   StockOptionLevel         4410 non-null   Int64  
 8   TrainingTimesLastYear    4410 non-null   Int64  
 9   YearsAtCompany           4410 non-null   Int64  
 10  YearsSinceLastPromotion  4410 non-null   Int64  
 11  YearsWithCurrManager     4410 non-null   Int64  
 12  TotalWorkingYears        4410 non-null   float64
 13  EnvironmentSatisfaction  4410 non-null   float64
 14  JobSatisfaction         

# 3. Análisis de Variables Categóricas

### 3.1. ¿Cuál fue el porcentaje de renuncias durante el año 2015?

In [5]:
funciones.univariado_torta(data, 'Attrition')

Attrition  Frecuencia Absoluta Frecuencia Relativa
0        No                 3825            86.735 %
1        Sí                  585            13.265 %

Aproximadamente el 13% de los colaboradores renunciaron en el año 2015.

### 3.2. ¿Cuál es el departamento que presentó el porcentaje más alto de retiros durante el año 2015?

In [6]:
funciones.analisisBivariado(data,['Attrition','Department'],'v','group')

Tabla de contingencia:


Department  Human Resources  Research & Development     Sales
Attrition                                                    
No                 0.036863                0.655163  0.307974
Sí                 0.082051                0.644444  0.273504

De los colaboradores que se retiraron, el 64.44% pertenecen al departamento de Investigación y Desarrollo. Además, de los colaboradores de Recursos Humanos, la proporción de colaboradores que se retiraron es el doble que la proporción de los que no se retiraron. En términos generales, el departamento con mayor número de retiros es Investigación y Desarrollo, pero hay que revisar Recursos Humanos porque el porcentaje de representación de los que se retiran es mayor que los que se quedan.

### 3.3. ¿Los retiros están relacionados con la frecuencia con la que los colaboradores deben viajar?

In [7]:
funciones.analisisBivariado(data, ['Attrition', 'BusinessTravel'], 'v','relative')

Tabla de contingencia:


BusinessTravel  Non-Travel  Travel_Frequently  Travel_Rarely
Attrition                                                   
No                0.109542           0.172026       0.718431
Sí                0.052991           0.295726       0.651282

De los colaboradores que se retiraron en el 2015, hubo más presencia de colaboradores que debían viajar frecuentemente. Además, de los colaboradores que debían viajar frecuentemente, la proporción de colaboradores que no se retiraron menor a la proporción de colaboradores que sí se retiraron, por lo que esta variable puede ser importante para predecir el retiro de un empleado.

### 3.4. ¿Qué influencia tiene el nivel de educación en el porcentaje de retiros?

In [8]:
funciones.analisisBivariado(data, ['Attrition', 'Education'], 'v', 'relative')

Tabla de contingencia:


Education  Bachelor  Below College   College    Doctor    Master
Attrition                                                       
No         0.389804       0.117386  0.186667  0.033725  0.272418
Sí         0.384615       0.104274  0.225641  0.025641  0.259829

No hay una influencia clara del nivel de educación con los retiros.

### 3.5. ¿Cuáles son los campos de estudio con más retiros?

In [9]:
funciones.analisisBivariado(data, ['Attrition', 'EducationField'], 'v', 'relative')

Tabla de contingencia:


EducationField  Human Resources  Life Sciences  Marketing   Medical     Other  \
Attrition                                                                       
No                     0.013856       0.411242   0.108497  0.316078  0.057516   
Sí                     0.047863       0.418803   0.105983  0.312821  0.044444   

EducationField  Technical Degree  
Attrition                         
No                      0.092810  
Sí                      0.070085

De los que se retiran, el porcentaje de colaboradores de Recursos Humanos es mayor en comparación con el porcentaje de colaboradores que no se retiraron siendod e Recursos Humanos.

### 3.6. Género

In [10]:
funciones.analisisBivariado(data,['Attrition','Gender'],"h","group")

Tabla de contingencia:


Gender       Female      Male
Attrition                    
No         0.402092  0.597908
Sí         0.386325  0.613675

No hay una dominancia de género en los retiros.

### 3.7. Rol del empleo

In [11]:
funciones.analisisBivariado(data,['Attrition','JobRole'],"v","relative")

Tabla de contingencia:


JobRole    Healthcare Representative  Human Resources  Laboratory Technician  \
Attrition                                                                      
No                          0.090458         0.035817               0.175686   
Sí                          0.080342         0.032479               0.179487   

JobRole     Manager  Manufacturing Director  Research Director  \
Attrition                                                        
No         0.072157                0.102745           0.051765   
Sí         0.051282                0.071795           0.071795   

JobRole    Research Scientist  Sales Executive  Sales Representative  
Attrition                                                             
No                   0.195033         0.218824              0.057516  
Sí                   0.222222         0.241026              0.049573

Solo el puesto de investigador científico tiene una mayor representación en los colaboradores que se retiran comparado con los que no.

### 3.8. Estado Civil

In [12]:
funciones.analisisBivariado(data,['Attrition','MaritalStatus'],"v","group")

Tabla de contingencia:


MaritalStatus  Divorced   Married    Single
Attrition                                  
No             0.234248  0.473464  0.292288
Sí             0.145299  0.355556  0.499145

Los colaboradores solteros tienen una mayor tendencia a retirarse de la empresa en comparación con los colaboradores casados o divorsiados.

# 4. Análisis de Satisfacción del Empleado

### 4.1. Satisfacción con el ambiente

In [13]:
funciones.analisisBivariado(data,['Attrition','EnvironmentSatisfaction'],"v","group")

Tabla de contingencia:


EnvironmentSatisfaction       1.0       2.0       3.0       4.0
Attrition                                                      
No                       0.176209  0.196601  0.317647  0.309542
Sí                       0.292308  0.177778  0.273504  0.256410

De los colaboradores que se retiraron de la empresa, la proporción que manifestaron una baja satisfacción con el ambiente es bastante mayor en comparación con los que no se retiraron. Por otro lado, los colaboradores que se quedaron presentaron una tendencia a manifestar una muy buena y excelente satisfacción con el ambiente.

### 4.2. Satisfacción con el empleo

In [14]:
funciones.analisisBivariado(data,['Attrition','JobSatisfaction'],"v","group")

Tabla de contingencia:


JobSatisfaction       1.0       2.0       3.0       4.0
Attrition                                              
No               0.182745  0.188497  0.297778  0.330980
Sí               0.275214  0.203419  0.314530  0.206838

Al igual que con la satisfacción con el entorno, la satisfacción con el empleo presenta calificaciones bajas por parte de los colaboradores que se retiraron de la empresa y mejores puntajes por parte de los que no se retiraron.

### 4.3. Balance entre vida y trabajo

In [15]:
funciones.analisisBivariado(data,['Attrition','WorkLifeBalance'],"v","group")

Tabla de contingencia:


WorkLifeBalance       1.0       2.0       3.0       4.0
Attrition                                              
No               0.044967  0.231895  0.621438  0.101699
Sí               0.114530  0.225641  0.548718  0.111111

EL balance entre vida y trabajo presenta calificaciones buenas tanto para colaboradores que se retiraron como para los que no. Aunque los que abandonaron tendieron un poco más a calificar esta variable como mala, hay una concentración alta en una buena calificación y poco en la mejor calificación.

# 5. Análisis de Desempeño del Empleado

### 5.1. Compromiso con el empleo

In [16]:
funciones.analisisBivariado(data,['Attrition','JobInvolvement'],"v","group")

Tabla de contingencia:


JobInvolvement         1         2         3         4
Attrition                                             
No              0.052026  0.256732  0.595817  0.095425
Sí              0.085470  0.244444  0.555556  0.114530

No hay una diferencia clara que permita concluir si esta variable influye en el retiro de los colaboradores.

In [17]:
funciones.analisisBivariado(data, ['retirementType', 'JobInvolvement'], 'v', 'relative')

Tabla de contingencia:


JobInvolvement         1         2         3         4
retirementType                                        
Fired           0.015385  0.353846  0.538462  0.092308
Resignation     0.085470  0.244444  0.555556  0.114530

No hay una indicación clara de que la calificación del compromiso con el empleo fuera una variable que influye con el retiro de los colaboradores y tampoco diferencia en tre los despedidos y los que renunciaron.

### 5.2. Calificación de Rendimiento

In [18]:
funciones.analisisBivariado(data,['Attrition','PerformanceRating'],"v","group")

Tabla de contingencia:


PerformanceRating         3         4
Attrition                            
No                 0.847320  0.152680
Sí                 0.839316  0.160684

El puntaje de rendimiento está casi que igualmente distribuido para ambos grupos, los que se retiraron y los que no.

In [19]:
funciones.analisisBivariado(data, ['retirementType', 'PerformanceRating'], 'v', 'relative')

Tabla de contingencia:


PerformanceRating         3         4
retirementType                       
Fired              0.784615  0.215385
Resignation        0.839316  0.160684

Se esperaría que los colaboradores despedidos presentara una mayor proporción en puntajes de rendimiento bajos, pero fue todo lo contrario.

Las variables de rendimiento, que son las que se esperaría que pudiesen explicar los despidos de colaboradores no lo están haciendo. Por lo que se opta por trabajar con algoritmos de cliasificación binaria, donde la salida sea si el colaborador se va o no.

# 6. Análisis de variables numéricas

### 6.1. Teniendo en cuenta los retiros ¿Varía de alguna durante cada mes del año? 

In [63]:
data['retirementDate'] = pd.to_datetime(data['retirementDate'])
data['mes_renuncia']=data['retirementDate'].dt.strftime('%B')

In [64]:
funciones.univariado_barras(data, 'mes_renuncia')

mes_renuncia  Frecuencia Absoluta Frecuencia Relativa
8           May                   69             1.565 %
4       January                   66             1.497 %
5          July                   63             1.429 %
3      February                   61             1.383 %
7         March                   59             1.338 %
9      November                   53             1.202 %
10      October                   53             1.202 %
6          June                   52             1.179 %
0         April                   48             1.088 %
1        August                   48             1.088 %
11    September                   46             1.043 %
2      December                   32             0.726 %

No se evidencia alguna variación tendencia o variación durante el año

### 6.2. Edad

In [65]:
funciones.bivariado_num(data,['Attrition', 'Age'])

 Edad del colaborador de los que no se retiraron: 
count       3825.0
mean     37.421176
std       8.944951
min           18.0
25%           31.0
50%           36.0
75%           43.0
max           60.0
Name: Age, dtype: Float64

 Edad del colaborador de los que si se retiraron: 
count        585.0
mean     33.671795
std       9.676885
min           18.0
25%           28.0
50%           31.0
75%           40.0
max           58.0
Name: Age, dtype: Float64


### 6.3. Distancia hogar a trabajo

In [66]:
funciones.bivariado_num(data,['Attrition', 'DistanceFromHome'])

 Distancia del domicilio en kms de los que no se retiraron: 
count      3825.0
mean     9.180915
std      8.131902
min           1.0
25%           2.0
50%           7.0
75%          14.0
max          29.0
Name: DistanceFromHome, dtype: Float64

 Distancia del domicilio en kms de los que si se retiraron: 
count       585.0
mean     9.268376
std      7.933463
min           1.0
25%           2.0
50%           7.0
75%          15.0
max          29.0
Name: DistanceFromHome, dtype: Float64


### 6.4. Ingresos mensuales de los trabajadores

In [67]:
funciones.bivariado_num(data,['Attrition', 'MonthlyIncome'])

 Ingresos mensuales en dólares al mes de los que no se retiraron: 
count          3825.0
mean     65437.416993
std      47320.247481
min           10510.0
25%           29350.0
50%           49070.0
75%           85000.0
max          199990.0
Name: MonthlyIncome, dtype: Float64

 Ingresos mensuales en dólares al mes de los que si se retiraron: 
count           585.0
mean     62360.940171
std      45339.795392
min           10090.0
25%           28440.0
50%           49500.0
75%           71400.0
max          198590.0
Name: MonthlyIncome, dtype: Float64


### 6.5. Número de compañías en las que ha trabajado

In [68]:
funciones.bivariado_num(data,['Attrition', 'NumCompaniesWorked'])

 Número total de empresas en las que ha trabajado el colaborador de los que no se retiraron: 
count    3825.000000
mean        2.643922
std         2.461791
min         0.000000
25%         1.000000
50%         2.000000
75%         4.000000
max         9.000000
Name: NumCompaniesWorked, dtype: float64

 Número total de empresas en las que ha trabajado el colaborador de los que si se retiraron: 
count    585.000000
mean       2.972650
std        2.693715
min        0.000000
25%        1.000000
50%        1.000000
75%        5.000000
max        9.000000
Name: NumCompaniesWorked, dtype: float64


### 6.6. Porcentaje del aumento salarial en el último año

In [69]:
funciones.bivariado_num(data,['Attrition', 'PercentSalaryHike'])

 Porcentaje de aumento salarial en el último año de los que no se retiraron: 
count       3825.0
mean     15.186667
std       3.656996
min           11.0
25%           12.0
50%           14.0
75%           18.0
max           25.0
Name: PercentSalaryHike, dtype: Float64

 Porcentaje de aumento salarial en el último año de los que si se retiraron: 
count        585.0
mean     15.358974
std       3.672513
min           11.0
25%           12.0
50%           14.0
75%           18.0
max           25.0
Name: PercentSalaryHike, dtype: Float64


### 6.7. Número de veces que se capacitó el trabajador

In [70]:
funciones.bivariado_num(data,['Attrition', 'TrainingTimesLastYear'])

 Número de veces que se impartió formación a este colaborador el año pasado de los que no se retiraron: 
count      3825.0
mean     2.825098
std      1.310761
min           0.0
25%           2.0
50%           3.0
75%           3.0
max           6.0
Name: TrainingTimesLastYear, dtype: Float64

 Número de veces que se impartió formación a este colaborador el año pasado de los que si se retiraron: 
count       585.0
mean     2.630769
std      1.122807
min           0.0
25%           2.0
50%           3.0
75%           3.0
max           6.0
Name: TrainingTimesLastYear, dtype: Float64


### 6.8. Años en la compañía

In [72]:
funciones.bivariado_num(data,['Attrition', 'YearsAtCompany'])

 Número total de años que el colaborador lleva en la empresa de los que no se retiraron: 
count      3825.0
mean      7.28915
std      6.137166
min           0.0
25%           3.0
50%           5.0
75%          10.0
max          40.0
Name: YearsAtCompany, dtype: Float64

 Número total de años que el colaborador lleva en la empresa de los que si se retiraron: 
count       585.0
mean      5.17094
std      5.719904
min           0.0
25%           1.0
50%           3.0
75%           7.0
max          40.0
Name: YearsAtCompany, dtype: Float64


### 6.9. Años desde la última promoción

In [74]:
funciones.bivariado_num(data,['Attrition', 'YearsSinceLastPromotion'])

 Número de años desde el último ascenso de los que no se retiraron: 
count      3825.0
mean     2.215425
std      3.234494
min           0.0
25%           0.0
50%           1.0
75%           3.0
max          15.0
Name: YearsSinceLastPromotion, dtype: Float64

 Número de años desde el último ascenso de los que si se retiraron: 
count       585.0
mean     2.006838
std       3.13344
min           0.0
25%           0.0
50%           1.0
75%           2.0
max          15.0
Name: YearsSinceLastPromotion, dtype: Float64


### 6.10. Años con el último jefe

In [75]:
funciones.bivariado_num(data,['Attrition', 'YearsWithCurrManager'])

 Número de años bajo el mando del jefe actual de los que no se retiraron: 
count      3825.0
mean     4.307712
std      3.589808
min           0.0
25%           2.0
50%           3.0
75%           7.0
max          17.0
Name: YearsWithCurrManager, dtype: Float64

 Número de años bajo el mando del jefe actual de los que si se retiraron: 
count       585.0
mean     2.916239
std      3.164144
min           0.0
25%           0.0
50%           2.0
75%           5.0
max          14.0
Name: YearsWithCurrManager, dtype: Float64


### 6.11. Total de años trabajando

In [76]:
funciones.bivariado_num(data,['Attrition', 'TotalWorkingYears'])

 Número total de años que el colaborador ha trabajado hasta ahora de los que no se retiraron: 
count    3825.000000
mean       11.732549
std         7.774701
min         0.000000
25%         6.000000
50%        10.000000
75%        16.000000
max        40.000000
Name: TotalWorkingYears, dtype: float64

 Número total de años que el colaborador ha trabajado hasta ahora de los que si se retiraron: 
count    585.000000
mean       8.300855
std        7.091748
min        0.000000
25%        3.000000
50%        7.000000
75%       10.000000
max       40.000000
Name: TotalWorkingYears, dtype: float64
